### DISCUSSION TITLE
Adding normals to a points cloud (for Poisson reconstruction)

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#6218] [Q&A] [answered]
Hello,

  Thanks to @sloriot I finally managed to do my first CGAL Poisson reconstruction. Here is my Poisson reconstructed solid Möbius strip:
![mobiussolidwire](https://user-images.githubusercontent.com/4466543/148122261-7a58bb55-bf12-48d6-9363-b030e006dccc.png)

Well, I was wondering something: if you don't have some normals, how do you compute them? I'm personally using the R package **Rvcg** to do that, a wrapper of the VCG C++ library. Is there such a functionality in CGAL? That would be nice.

**Comment by sloriot** [comment, marked as answer]
You can use [CGAL::jet_estimate_normals()](https://doc.cgal.org/latest/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga0cd0f87de690d4edf82740e856efa491) to estimate normal directions followed by 
[CGAL::mst_orient_normals()](https://doc.cgal.org/latest/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga50c98d5c5ae5535bce6f32eddbd03f33) to estimate normal orientations

**Comment by stla** [comment]
Thanks! I found [this example](https://github.com/CGAL/cgal/blob/master/Point_set_processing_3/examples/Point_set_processing_3/normal_estimation.cpp) and I'm currently doing the implementation. At first glance the results look quite different from the ones of (R)vcg.

**Comment by stla** [comment]
It works! Here is the Poisson reconstructed solid Möbius strip with the VCG normals and with the CGAL normals. Hard to see any difference.

![mobiussolid_vcgnormals](https://user-images.githubusercontent.com/4466543/148399415-4a04a5f2-0bd7-4e99-b6ab-1e09cb299490.png)

![mobiussolid_cgalnormals](https://user-images.githubusercontent.com/4466543/148399453-01d9fdf8-bd37-42aa-acc5-4960b4dc48fc.png)

**Reply by sloriot** [reply]
Could you share the meshes? It seems that the one generated using CGAL normals is less smooth on the bottom part of the picture. If you also have the points set with normals if might be interesting to compare (You can use [that function](https://doc.cgal.org/latest/Point_set_processing_3/group__PkgPointSetProcessing3IO.html#ga93f0ee77f3b2a1bd7dc70482a4e349fc)).

**Comment by stla** [comment]
Hello @sloriot ,

  Which format do you want for the meshes?

**Reply by sloriot** [reply]
what ever works for you. The easier is `std::ofstream("out.off") << std::setprecision(17) << poly; `

**Comment by stla** [comment]
Ok, will try. Here is another example. Indeed, the mesh is smoother with the VCG normals (yellow).

![toroidahelix_vvg](https://user-images.githubusercontent.com/4466543/148532684-ad1b0364-e546-4b8d-99b2-62b6b457a686.png)

![toroidahelix_cgal](https://user-images.githubusercontent.com/4466543/148532753-a4c44ee7-4d20-4914-b215-ac2d48dc9ccd.png)

**Comment by stla** [comment]
Here are the files. These are the files for the toroidal helix, not for the solid Möbius strip, I choosed these ones because the difference is more striking.

**NB:** The `off.txt` files actually are `off` files. Github does not accept them, so I added the `txt` suffix.

[original_mesh.txt](https://github.com/CGAL/cgal/files/7828628/original_mesh.txt)
[out_VCGnormals.off.txt](https://github.com/CGAL/cgal/files/7828635/out_VCGnormals.off.txt)
[out_CGALnormals.off.txt](https://github.com/CGAL/cgal/files/7828636/out_CGALnormals.off.txt)

**Reply by afabri** [reply]
I think it would  be better to have a look at the point cloud itself.

**Reply by sloriot** [reply]
it is in original_mesh.txt

**Reply by sloriot** [reply]
together with CGAL and VCG normals

**Reply by sloriot** [reply]
will shortly create a gist with files readable in the polyhedron demo

**Reply by sloriot** [reply]
https://gist.github.com/sloriot/ed7df61ca280a003ce0896458df7fb16

**Reply by sloriot** [reply]
Thanks for the file!
